watchers/disruptions_watchers_manager.go

```diff
@@ -0,0 +1,217 @@
+	// CreateAllWatchers adds new Watchers instances for a disruption
+	CreateAllWatchers(disruption *v1beta1.Disruption, watcherManagerMock Manager, cacheMock k8scontrollercache.Cache) error
+
+	// RemoveAllWatchers removes all existing Watchers of a disruption
+	RemoveAllWatchers(disruption *v1beta1.Disruption)
+
+	// RemoveAllOrphanWatchers removes all watchers linked to an expired disruption
+	RemoveAllOrphanWatchers() error
+
+	// RemoveAllExpiredWatchers removes all expired Watchers
+	RemoveAllExpiredWatchers()
+}
+
+// WatcherManagers represents a map of Manager instances
+type WatcherManagers map[types.NamespacedName]Manager
+
+// disruptionsWatchersManager is the struct that implement the DisruptionsWatchersManager interface.
+type disruptionsWatchersManager struct {
+	controller       controller.Controller
+	factory          Factory
+	log              *zap.SugaredLogger
+	reader           client.Reader
+	watchersManagers WatcherManagers
+}
+
+type WatcherName string
+
+const (
+	ChaosPodWatcherName         WatcherName = "ChaosPod"
+	DisruptionTargetWatcherName WatcherName = "DisruptionTarget"
+)
+
+var watchersNames = []WatcherName{
+	ChaosPodWatcherName,
+	DisruptionTargetWatcherName,
+}
+
+// CreateAllWatchers creates all the Watchers associated with the given Disruption.
+func (d disruptionsWatchersManager) CreateAllWatchers(disruption *v1beta1.Disruption, watcherManagerMock Manager, cacheMock k8scontrollercache.Cache) error {
+	// Check that the disruption object has a name and namespace
+	if disruption.ObjectMeta.Name == "" || disruption.ObjectMeta.Namespace == "" {
+		return fmt.Errorf("the disruption is not valid. It should contain a name and a namespace")
+	}
+
+	// Get the namespaced name of the disruption
+	disruptionNamespacedName := getDisruptionNamespacedName(disruption)
+
+	var watcherManager Manager
+
+	// If a mock watcher manager was passed in, use it
+	if watcherManagerMock == nil {
+		watcherManager = d.getWatcherManager(disruptionNamespacedName)
+	} else {
+		watcherManager = watcherManagerMock
+	}
+
+	// Save the watcher manager for later use
+	d.watchersManagers[disruptionNamespacedName] = watcherManager
+
+	// Calculate a hash of the disruption spec (excluding the count field)
+	disSpecHash, err := disruption.Spec.HashNoCount()
+	if err != nil {
+		return err
+	}
+
+	// For each type of watcher we need to create
+	for _, watcherName := range watchersNames {
+		// Calculate a hash for the watcher
+		watcherNameHash := disSpecHash + string(watcherName)
+
+		// If a watcher with this hash has already been created
+		if ok := watcherManager.GetWatcher(watcherNameHash); ok != nil {
+			continue
+		}
+
+		// Otherwise add the new watcher for the disruption
+		if err := d.addWatcher(disruption, watcherName, watcherNameHash, cacheMock, watcherManager); err != nil {
+			return err
+		}
+
+		d.log.Infof("Watcher %s created for disruption %s with %s namespace", watcherName, disruption.Name, disruption.Namespace)
```

Comment:
This should be using the `disruptionName` and `disruptionNamespace` tags, rather than embedding them in the string. I also think it should be ok for this to be at debug?

Reply:
ditto on the other places in this file where we use Infof instead of Infow